src/twisted/internet/test/test_tcp.py

```diff
@@ -1230,12 +1230,26 @@ def stopReactorIfServerAccepted(_):
 
     serverConnectionMade.addCallback(stopReactorIfServerAccepted)
 
-    port = listen(reactor, serverFactory)
-    listeningHost = port.getHost()
     clientFactory = MyClientFactory()
-    connect(reactor, listeningHost, clientFactory)
 
-    reactor.callWhenRunning(exhauster.exhaust)
+    if hasattr(reactor, "callLater"):
+        # For Glib-based reactors, the exhauster should be run after the signal handler used by glib [1]
+        # and restoration of twisted signal handlers [2], so I use callLater with non-zero time
+        # [1] https://gitlab.gnome.org/GNOME/pygobject/-/blob/3.42.0/gi/_ossighelper.py#L76
+        # [2] https://github.com/twisted/twisted/blob/twisted-22.4.0/src/twisted/internet/_glibbase.py#L134
+        def inner():
+            port = listen(reactor, serverFactory)
+            listeningHost = port.getHost()
+            connect(reactor, listeningHost, clientFactory)
+            exhauster.exhaust()
+
+        reactor.callLater(0.01, inner)
```

Comment:
Given the implementation details of `_glibbase.py` I think it might be safer to do `reactor.callLater(0, reactor.callLater, 0, inner)`.

Depending on how time passes and the exact scheduling of the thread running the reactor in this test, `callLater(0.01, ...` could end up being equivalent to `callLater(0, ...` which I suppose would still potentially interact poorly with `_globbase.py`'s re-installation of signal handlers.  The double `callLater(0, ...` should avoid this because no matter how time passes, the first `callLater(0, ...` here and the `callLater(0, ...` in `_glibbase.py` will happen in the same reactor iteration and only after that can the second `callLater(0, ...` here happen.

Reply:
That makes sense. The `reactor.callLater(0, reactor.callLater, 0, inner)` trick is brilliant :)